docs: 📚 add developer starting guide (#680)
solid-design-system/solid: DEVELOPER_STARTING_GUIDE.md

  - [slots](https://developer.mozilla.org/en-US/docs/Web/HTML/Global_attributes/slot) allow for easy element nesting ([Lit: Working with Shadow DOM](https://lit.dev/docs/components/shadow-dom/))
  - [parts](https://developer.mozilla.org/en-US/docs/Web/HTML/Global_attributes/part) allow external manipulation of CSS.
  - [events](https://lit.dev/docs/components/events/) make critical information available outside of a component.

### Samples & Patterns

Alongside the documentation for Styles and Components, we also present Samples and Patterns. These differ in their scope:

- **Samples**: Tied to a specific component, these show its use in a particular context. They are added as "Sample" stories in the component's documentation.
- **Patterns**: More advanced, they demonstrate how to combine several components to solve a specific problem. They are added as stories in the "Patterns" section.

## Shoelace

The [Shoelace](https://shoelace.style/) web component library is our working inspiration. Components are copied directly from their [source code](https://github.com/shoelace-style/shoelace) and pasted in the `packages/components/src/_components` folder.

> _There may have been updates since the files there were copied to our repo!_

Reviewer: This statement might need some followup instructions. New devs might not infer that they should copy the components from the Shoelace repo in these cases instead of relying on the ones in _components.

Something along the lines of "In case of updates, copy the source code from the latest version of the Shoelace repository."

Author: Get your point. 
Any suggestions?

Reviewer: Added a suggestion below so closing this to avoid redundancy.

> _There may have been updates since the files there were copied to our repo!_

Reviewer: ```suggestion
_There may have been updates since the files there were copied to our repo!_ In those instances, copy the updated component folder from the Shoelace repository and use that as a starting point.
```

12. **Add any tests that are needed for new component behavior.** Motto: “Test the behavior, not the implementation.” Write tests that verify your components meet the expected requirements and specifications, ensuring they function correctly from the user's perspective. Don’t get too tied up trying to test all the technical details.
13. **Author the full migration guide.** The migration guide should be placed in the `packages/components/src/docs/Migration` folder. The migration guide should be based on the migration guide template (`templates/migration-guide-template.mdx`) and be named by the old component name (e.g. ui-button.mdx).

**Run `pnpm verify` in the root directory before creating a PR.** This will check that all formatting, tests, and build processes are working correctly to allow the pipeline to run successfully.

### Styles

If you are working on a style, follow the steps below:

> ...

Reviewer: This steps might have been skipped or missing.

Author: Right... just came to my mind, that this part needs to be defined. 

@Vahid1919 do you have any suggestions how to write/explain the steps?

Reviewer: Yes. I will look through the styles I've done before and write some sort of draft.

Reviewer: ```suggestion
1. Create new styles in the `packages/components/src/styles` folder.
2. Set up a Storybook story alongside the styles, using a default story. This allows you to preview and iterate on the styles as you work.
3. **Each style should have an accompanying declaration file.** This is for documenting the style and its attributes. Follow the structure of our existing styles for guidance (eg: [sd-meta](https://github.com/solid-design-system/solid/blob/main/packages/components/src/styles/meta/meta.declaration.ts)).
4. Import your styles into the `packages/components/src/solid-styles.css` file. This step is crucial for enabling other Solid components to access the new styles.

_ You can always refer to the existing components and styles to familiarize yourself with the workflow!_
```

Reviewer: @karlbaumhauer I think this covers it when it comes to making styles. Feel free to edit and reword as you see fit :)

        - `accordion`: Lit JS web component that implements an accordion. The comments are automatically compiled into the Storybook docs and should be used for user clarification.
        - `accoridion.tests.ts`: Suite of jest tests for the accordion component. Critical for deployment and maintenance.
        - `accordion.stories.ts`: Collection of Storybook stories that are primarily used for visual testing with Chromatic. Any **samples** are added as stories here.
      - `docs`: all ".mdx" doc files that appear in the "DOCS" section of our Storybook
        - `migration`: Individual component migration guides
      - `internal`: A set of utilities that are reusable internally. Check here should you need something like a debounce, it may have already been implemented.
      - **`patterns`**: Demonstrate how to combine several components to solve a specific problem. Added as stories in the "PATTERNS" section.
      - **`styles`**: These are standalone CSS files. They don't provide any logic and are often non-interactive. Added as stories in the "STYLES" section.

## Technologies

- pnpm
- Web Components
- Lit JS
- Tailwind
- Storybook
- Jest
- Chromatic
- Vite
- PostCSS

> //TBD: Do we keep them and apply them even thought the nome does not come from the SOLID principles?

Reviewer: comment accidentally left in?

Author: Left here on purpose for discussion. Especially for @yoezlem and @mariohamann...;)

Reviewer 2: I would like to keep it. Both the principles with their foundations and the name fit perfectly with us. This guide is written for developers who are not familiar with our stack, our way of working, and our principles. For me as a new developer, it would be helpful.

Reviewer 2: I would like to keep it. Both the principles with their foundations and the name fit perfectly with us. This guide is written for developers who are not familiar with our stack, our way of working, and our principles. For me as a new developer, it would be helpful.

Author: ```suggestion
```

> ## SOLID Design Principles in Software Development
> 
> [SOLID](https://www.freecodecamp.org/news/solid-design-principles-in-software-development/) is a set of five design principles that aid in designing robust, testable, extensible, and maintainable object-oriented software systems. Each principle addresses specific challenges in software development.
> 
> - **Single Responsibility Principle (SRP)**
> 
>   - A class, module, or function should have only one reason to change, i.e., it should perform a single responsibility.
>   - Example: Separating classes for animal information, sound, and feeding.
> 
> - **Open-Closed Principle (OCP)**
> 
>   - Classes, modules, and functions should be open for extension but closed for modification.
>   - Example: Using abstraction to add new animal types without modifying existing code.
> 
> - **Liskov Substitution Principle (LSP)**
> 
>   - Child classes must be substitutable for their parent classes, ensuring seamless replacement.
>   - Example: Dog and Cat classes replacing the parent Animal class.
> 
> - **Interface Segregation Principle (ISP)**
> 
>   - Clients should not be forced to implement interfaces or methods they do not use.
>   - Example: Breaking down large interfaces into smaller, specific ones.
> 
> - **Dependency Inversion Principle (DIP)**
>   - High-level modules should not depend on low-level modules; both should depend on abstractions.
>   - Example: Using abstractions to decouple high-level and low-level modules.
> 
> Implementing SOLID principles leads to a large but maintainable codebase, enabling developers to make changes without causing major issues. These principles contribute to bug-free, flexible, scalable, and reusable software systems.

Author: ```suggestion
## SOLID Design Principles in Software Development

After defining our own principles and decided for the name Solid Design System, we found the [SOLID Design Principles](https://www.freecodecamp.org/news/solid-design-principles-in-software-development/). A set of five design principles that aid in designing robust, testable, extensible, and maintainable object-oriented software systems. Each principle addresses specific challenges in software development.

These principles are not developed by us and the connection is more 'unintentional'. Nevertheless, the strong and deep connection between our principles and the ones mentioned above is definitely worth sharing here and we highly recommend to get familiar with them as well.
```

> Implementing SOLID principles leads to a large but maintainable codebase, enabling developers to make changes without causing major issues. These principles contribute to bug-free, flexible, scalable, and reusable software systems.

Reviewer 3: ```suggestion
## SOLID Design Principles in Software Development

After defining our own principles and decided for the name Solid Design System, we realized the close connection to [SOLID Design Principles](https://www.freecodecamp.org/news/solid-design-principles-in-software-development/), a set of five design principles that aid in designing robust, testable, extensible, and maintainable object-oriented software systems. Each principle addresses specific challenges in software development.
These principles are not developed by us and the connection is more 'unintentional'. Nevertheless, the strong and deep connection between our principles and the ones mentioned above is definitely worth sharing here and we highly recommend to get familiar with them as well.
```

I just a little bit the wording – so that it doesn't sound like we never heard of it.  :)

## Shoelace

The [Shoelace](https://shoelace.style/) web component library is our working inspiration. Components are copied directly from their [source code](https://github.com/shoelace-style/shoelace) and pasted in the `packages/components/src/_components` folder.

_There may have been updates since the files there were copied to our repo!_ In those instances, copy the updated component folder from the Shoelace repository and use that as a starting point.

Additionally we have taken things like the `FormControlController` directly from Shoelace and are slowly adapting them to our needs. If something is broken, missing, or working in an unexpected manner, there may be things that were missed / misconfigured / not yet addressed in our library.

A good rule of thumb: **when in doubt, refer to Shoelace!**

## Storybook

We present the library in a [Storybook](https://storybook.js.org/) with the following sections:

- **DOCS**: All .mdx documentation files.
- **COMPONENTS**: Web Components built with Lit JS. Often more complex than Styles, and could feature reactivity, state, multiple slots, properties, and more.

> - **UTILITIES**:

Reviewer 2: The description is missing here.. Maybe something like this:
UTILITIES: Services that provide essential functionality.

Author: ```suggestion
- **UTILITIES**: Web Components built with Lit JS, used as SDS-internal helper components. Technically necessary to build other components and generic enough to be used multiple times. Not officially part of the design system library and no respective component existing in Figma.
```

Author: what do you think @yoezlem ?

4. **Remove any properties from the copied Shoelace code that are irrelevant to our design** (e.g. `pill` is a re-occurring prop in Shoelace that alters a component's border radius, our designs do not offer this). Refer to the [Figma design documentation](https://www.figma.com/files/1075429990769806468/project/67503549/Solid-DS-Documentation?fuid=883643809929820461) for the component you are working on.
5. **Compare the old UI component to the newly adapted Shoelace component.** Refer to the [UI Component Library](https://component-library.dev.fe.union-investment.de/integration/storybook/index.html) to get an idea of what is being changed. This helps to see what is being lost or improved upon and should be summarized in the migration guide.
6. **Keep the technical props and features of the Shoelace component that feel relevant.** It is easier to decide which ones can be removed at the end.

> 7. **Convert the component styles to use a Tailwind first approach.** Shoelace does not use Tailwind and includes CSS files for every component. These should be carefully inspected and adapted. Work your way through the render() method of the component and remove CSS classes one by one, replacing them with a combination of Tailwind classes that implement our design. Keep in mind our designs may differ from the original styles applied by Shoelace, however, there are many small details (regarding `display`, `position`, or interaction states etc.) that should be carefully observed and adapted.

Reviewer 2: Maybe we should also mentioned here that our rule is 'mobile first'.  Developers should keep in mind to use TailwindCSS accordingly.

Author: Is this really still the case? isn't the device no longer a criteria for us as we don't want to care about where our components are used?

Reviewer 2: That's correct, we don't prioritize how and where the Solid components are integrated and for which use case they will ultimately be used. However, we still build our components following the Mobile First approach. For instance, "text-3xl lg:text-4xl." We need to do this to ensure that our components are displayed correctly on both desktop and mobile. In essence, we initially oriented ourselves this way, and it's essential to stay true to the principle to avoid mixed desktop/mobile-first approaches everywhere.

Author: added a suggestion below @yoezlem. What do you think?

Reviewer 2: LGTM

> 8. **Implement any missing behavior.** Once the Shoelace component has been reduced and re-styled with Tailwind, add any missing functions that may be necessary to meet the component's requirements.
> 9. **Write concise comments build up the auto-generated Storybook docs.** We rely primarily on the docs to communicate to the library users (developers). Write descriptive comments and organize properties semantically to the best of your ability.
> 10. **Write concise Storybook stories to cover all visual scenarios.** See the Storybook section above for further information. All stories can furthermore be observed together in the docs.
> 11. **Start with the jest tests from Shoelace.** Inside of `packages/components` you can test individual components with `pnpm test.component input` (where input is akin to the sd-input component). These will likely contain some misnamed properties or improper default values. Adapt them as needed to fit our use case.
> 12. **Add any tests that are needed for new component behavior.** Motto: “Test the behavior, not the implementation.” Write tests that verify your components meet the expected requirements and specifications, ensuring they function correctly from the user's perspective. Don’t get too tied up trying to test all the technical details.
> 13. **Author the full migration guide.** The migration guide should be placed in the `packages/components/src/docs/Migration` folder. The migration guide should be based on the migration guide template (`templates/migration-guide-template.mdx`) and be named by the old component name (e.g. ui-button.mdx).

Author: ```suggestion
8. **Mobile First Approach.** We as design system do not consider in detail on which device our components will be used as we want to be as generic as possible. However, we still build our components following the Mobile First approach (eg. using mobile tokens as default).
9. **Implement any missing behavior.** Once the Shoelace component has been reduced and re-styled with Tailwind, add any missing functions that may be necessary to meet the component's requirements.
10. **Write concise comments build up the auto-generated Storybook docs.** We rely primarily on the docs to communicate to the library users (developers). Write descriptive comments and organize properties semantically to the best of your ability.
11. **Write concise Storybook stories to cover all visual scenarios.** See the Storybook section above for further information. All stories can furthermore be observed together in the docs.
12. **Start with the jest tests from Shoelace.** Inside of `packages/components` you can test individual components with `pnpm test.component input` (where input is akin to the sd-input component). These will likely contain some misnamed properties or improper default values. Adapt them as needed to fit our use case.
13. **Add any tests that are needed for new component behavior.** Motto: “Test the behavior, not the implementation.” Write tests that verify your components meet the expected requirements and specifications, ensuring they function correctly from the user's perspective. Don’t get too tied up trying to test all the technical details.
14. **Author the full migration guide.** The migration guide should be placed in the `packages/components/src/docs/Migration` folder. The migration guide should be based on the migration guide template (`templates/migration-guide-template.mdx`) and be named by the old component name (e.g. ui-button.mdx).
```
